src/10-ipfs-0-4-0-released/index.md

```diff
@@ -0,0 +1,121 @@
+
+The same code that allows us to easily select a stream muxer is also being
+used to select which ipfs sub-protocol to use for any given stream between
+peers. Now, if we need to make a breaking protocol change to any of those (like
+the DHT or bitswap) we can do so seamlessly, and provide easy backwards
+compatibility.
+
+In addition to the multiplexers changes, the protocol revamp has also improved
+efficiency and performance in a few important ways, including the elimination of
+a double wrapping of the length prefixer, and the removal of some unneccessary
+round trips between nodes.
+
+## Changes to the repo
+
+We also made a few changes to our on disk storage format (called the 'repo').
+The way object pinning (`ipfs pin add`) works has also been upgraded to be much
+more efficient, which will improve the overall speed of adding and downloading
+IPFS data. Previously, when you pinned a file recursively, it would add a
+recursive type pin to the root, and then add an indirect type pin to each child
+node. This made enumeration for GC convenient, but was very slow and made a lot
+of things needlessly complicated. We have switched to just adding the recursive
+pin to the root object, and then doing the enumeration of child nodes when we
+actually run a garbage collection.
+
+Because of this change, you will need to run a migration (from repo version 2 to
+version 3). If you update with the `ipfs-update` tool, this will be done for you
+automatically.
+
+## Other improvements, fixes
+
+In addition to a major protocol improvement and upgrade, this release adds a lot
+of new functionality, performance speedups, and stability fixes that make this
+the best version of IPFS to date.
+
+This release also includes performance and usability improvements to IPNS, the
+IPFS way to do mutable data. IPNS uses keypair encryption to allow users to
+point a pubkeyhash to an ipfs object in a way that is crytographically
+verifiable. By allowing users to change what the pubkeyhash points to, this
+provides users with a single hash they can give to users to get the latest
+version of their data. This creates a seamless way to use IPFS to verify
+content, and to distribute content via trustless nodes in a smart, safe way.
+This brings IPFS closer to the goal of being a global filesystem of data, that
+can allow everyone in the world to help serve the world's data in a way that
+enriches and empowers everyone.
+
+## How to upgrade
+
+Depending on how you initially installed ipfs, there are several ways to
+upgrade. If you installed ipfs with a pre-built binary, you can either head over
+to [dist.ipfs.io](https://dist.ipfs.io/#go-ipfs) and grab the latest version
+from there. Or alternatively, from the same page you can grab the `ipfs-update`
+binary, and use it to perform the upgrade for you. If you installed from source,
+you can simply update your git repo to the `v0.4.0` tag and run `make install`.
```

Comment:
We might have to mention gx here -- I had to manually run `make gx_upgrade` and `make gxgo_upgrade` here the other day.